### PR TITLE
SNI - Tunnel: Add support to match captured group from fqdn into tunnel_route

### PR DESCRIPTION
**Pull request by brbzull0**
This feature allows us to use the captured string from a   wildcarded `fqdn`( `sni.yaml`) inside the `tunnel_route` field. This  is useful when need to white-list a bunch of `fqdn` independently, so instead of having to add them one by one every time, having this few functionality, we can now just define a wildcard(already supported) in the `fqdn` and just let it match with a specific group in the `tunnel_route`. Groups have a specific syntax  `$N` being `N` one of the groups in the `fqdn`.

Existing approach:
```
sni:
   - fqdn: 'd1.foo.com'
     tunnel_route: 'd1.example'
   - fqdn: 'd2.foo.com'
     tunnel_route: 'd2.example'
   - fqdn: 'd3.foo.com'
     tunnel_route: 'd3.example'
```
With the new feature:
```
sni:
   - fqdn: '*.foo.com'
     tunnel_route: '$1.example'
```
This change also support multiple groups which can be match independently in `tunnel_route`.

*This came as an internal request from one of the groups in Verizon Media*

### CODE REVIEW

**Comment by randall** [comment]
[approve ci autest]

File: iocore/net/P_SNIActionPerformer.h [hunk 1]
```diff
@@ -81,21 +92,98 @@ class ControlH2 : public ActionItem
...
+          real_dst += groups[group_index - 1];
+          // if it was the last match, then ...
+          if (dot == std::string::npos && to == std::string::npos) {
+            // that's it.
+            break;
+          }
+          pos += number_str.size() + 1;
+          std::advance(c, number_str.size() + 1);
+        }
+      }
```
Comment (shinrich) [inline iocore/net/P_SNIActionPerformer.h:179]: If there was not a matching capture group, do we lose the $# characters?

Comment (brbzull0) [reply iocore/net/P_SNIActionPerformer.h:179]: If no matching group, we keep the `$#`.

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Pull request merged**